File: cpp/String/L_detStrHalfAlike.cpp
#include <bits/stdc++.h>
using namespace std;

bool halvesAreAlike(string S)
{
    string vowels = "aeiouAEIOU";
    int mid = S.size() / 2, ans = 0;
    for (int i = 0, j = mid; i < mid; i++, j++)
    {
        if (~vowels.find(S[i]))
            ans++;
        if (~vowels.find(S[j]))
            ans--;
    }
    return ans == 0;
}

bool halvesAreAlike(string s)
{
    set<char> vow = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
    int n = s.size() / 2, f = 0;
    for (int i = 0; i < n; i++)
    {
        f += (vow.count(s[i]) - vow.count(s[n + i]));
    }
    return f == 0;
}

File: Python/Strings/L_detStrHalfAlike.py
def halvesAreAlike(s: str) -> bool:
        count = 0
        vowels = set("aeiouAEIOU")
        n = len(s) // 2
        for i, ch in enumerate(s):
            if ch in vowels:
                if i < n:
                    count += 1
                else:
                    count -= 1
        return count == 0
    
def halvesAreAlike1(s: str) -> bool:
    def vowel_count(half: str) -> int:
        vowels = "aeiouAEIOU"
        count = 0
        for ch in half:
            if ch in vowels:
                count +=1
        return count
    mid_point = len(s) // 2
    return vowel_count(s[:mid_point]) == vowel_count(s[mid_point:])

print((halvesAreAlike1("book")))